File: Problems/Heaps/Top-K-Frequent-Elements/Top-K-Frequent-Elements.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>

// Using a heap - Time complexity: O(nlogk)

class Solution {
public:
    std::vector<int> topKFrequent(std::vector<int> &nums, int k)
    {
        if (nums.size() == k)
            return nums;
        std::unordered_map<int, int> cnt;

        for (int n : nums)
            cnt[n]++;
        std::vector<int> ans;

        if (cnt.size() == k)
        {
            for (auto &[n, c] : cnt)
                ans.push_back(n);
            return ans;
        }

        auto cmp = [&](int a, int b)
        { return cnt[a] > cnt[b]; };
        std::priority_queue<int, std::vector<int>, decltype(cmp)> pq(cmp); // Min heap

        for (auto &[n, c] : cnt)
        {
            pq.push(n);
            if (pq.size() > k)
                pq.pop();
        }
        while (pq.size())
        {
            ans.push_back(pq.top());
            pq.pop();
        }

        ans = std::vector<int>(ans.rbegin(), ans.rend());
        return ans;
    }
};

int main()
{
    std::vector<int> nums = {1, 1, 1, 2, 2, 3};
    int k = 2;

    Solution sol;
    std::vector<int> result = sol.topKFrequent(nums, k);

    std::cout << "Input: nums = [";
    for (int i = 0; i < nums.size(); i++)
    {
        std::cout << nums[i];
        if (i < nums.size() - 1)
            std::cout << ", ";
    }
    std::cout << "], k = " << k << std::endl;

    std::cout << "Output: [";
    for (int i = 0; i < result.size(); ++i)
    {
        std::cout << result[i];
        if (i < result.size() - 1)
            std::cout << ", ";
    }
    std::cout << "]" << std::endl;

    return 0;
}

File: Problems/Heaps/Top-K-Frequent-Elements/Top-K-Frequent-Elements.cs
using System;
using System.Collections.Generic;

// Using a heap - Time complexity: O(n log k)

public class Solution
{
    public int[] TopKFrequent(int[] nums, int k)
    {
        if (nums.Length == k) return nums;
        var cnt = new Dictionary<int, int>();

        foreach (int n in nums)
        {
            if (cnt.ContainsKey(n)) cnt[n]++;
            else cnt.Add(n, 1);
        }

        List<int> ans = new List<int>();

        if (cnt.Count == k)
        {
            foreach (var item in cnt) ans.Add(item.Key);
            return ans.ToArray();
        }

        var cmp = Comparer<int>.Create((a, b) => cnt[a] != cnt[b] ? cnt[a].CompareTo(cnt[b]) : a.CompareTo(b));
        var ss = new SortedSet<int>(cmp);

        foreach (var item in cnt)
        {
            ss.Add(item.Key);
            if (ss.Count > k) ss.Remove(ss.Min);
        }
        while (ss.Count > 0)
        {
            ans.Add(ss.Min);
            ss.Remove(ss.Min);
        }

        ans.Reverse();
        return ans.ToArray();
    }
}

class Program
{
    static void Main()
    {
        int[] nums = { 1, 1, 1, 2, 2, 3 };
        int k = 2;

        Solution sol = new Solution();
        int[] result = sol.TopKFrequent(nums, k);

        Console.Write("Input: nums = [");
        Console.Write(string.Join(", ", nums));
        Console.WriteLine($"], k = {k}");
        Console.Write("Output: [");
        Console.Write(string.Join(", ", result));
        Console.WriteLine("]");
    }
}